core/src/banking_stage/latest_unprocessed_votes.rs

```diff
@@ -262,49 +262,41 @@ impl LatestUnprocessedVotes {
         bank: Arc<Bank>,
         forward_packet_batches_by_accounts: &mut ForwardPacketBatchesByAccounts,
     ) -> usize {
-        let mut continue_forwarding = true;
         let pubkeys_by_stake = weighted_random_order_by_stake(
             &bank,
             self.latest_votes_per_pubkey.read().unwrap().keys(),
         )
         .collect_vec();
-        pubkeys_by_stake
-            .into_iter()
-            .filter(|&pubkey| {
-                if !continue_forwarding {
-                    return false;
-                }
-                if let Some(lock) = self.get_entry(pubkey) {
-                    let mut vote = lock.write().unwrap();
-                    if !vote.is_vote_taken() && !vote.is_forwarded() {
-                        let deserialized_vote_packet = vote.vote.as_ref().unwrap().clone();
-                        if let Some(sanitized_vote_transaction) = deserialized_vote_packet
-                            .build_sanitized_transaction(
-                                &bank.feature_set,
-                                bank.vote_only_bank(),
-                                bank.as_ref(),
-                            )
-                        {
-                            if forward_packet_batches_by_accounts.try_add_packet(
-                                &sanitized_vote_transaction,
-                                deserialized_vote_packet,
-                                &bank.feature_set,
-                            ) {
-                                vote.forwarded = true;
-                            } else {
-                                // To match behavior of regular transactions we stop
-                                // forwarding votes as soon as one fails
-                                continue_forwarding = false;
-                            }
-                            return true;
+        let mut count: usize = 0;
+        for pubkey in pubkeys_by_stake {
```

Comment:
`for` statement should handle an iterator just fine.
Removing a need for a vector to be constructed.

```diff
-         .collect_vec();
          let mut count: usize = 0;
          for pubkey in pubkeys_by_stake {
```

Reply:
Yeah, I figured there might be an issue with lifetimes there.
But, I think, it is because the function time is not correctly inferred.
The returned iterator does not hold any references to the input data, so there should be no lifetimes it is capturing.

Try specifying explicitly that the returned object does not capture any lifetimes:

```rust
pub(crate) fn weighted_random_order_by_stake<'a>(
    bank: &Bank,
    pubkeys: impl Iterator<Item = &'a Pubkey>,
) -> impl Iterator<Item = Pubkey> + 'static {
```

I think `'a` gets added to the returned type for some reason, as if you said
```rust
... -> impl Iterator<Item = Pubkey> + 'a {
```

Reply:
At the same time, I think, it might be possible to reduce the number of copies, if it would actually return `Item = &'a Pubkey`.
In which case you would need to put `binding` on the stack and keep is there for the whole duration of the algorithm.
As the returned iterator would reference `binding`.
Depending on the lock contention, one or the other might be faster.